js/src/gc/RootMarking.cpp

```diff
@@ -48,7 +48,16 @@ using RootEnum = RootedValueMap::Enum;
 // ConcreteTraceable type are actually used at runtime.
 struct ConcreteTraceable {
   ConcreteTraceable() = delete;
-  void trace(JSTracer*) { MOZ_CRASH("This path is unreachable."); }
+#ifdef _MSC_VER
+  // There should be no callers to ConcreteTraceable::trace(). In clang/gcc, it
+  // is possible to ensure that using the "= delete" syntax, but MSVC is more
+  // aggressive about what is considered a caller.
+  void trace(JSTracer*) {
+    MOZ_ASSERT_UNREACHABLE();
+  }
+#else
+  void trace(JSTracer*) = delete;
+#endif
```

Comment:
```suggestion
void trace(JSTracer*) { MOZ_ASSERT_UNREACHABLE(); }
```

I'm pretty sure that `= delete` also leads to a compilation issue on GCC. Isn't it better to just use a default unreachable implementation?

Reply:
No it isn't better *IF* it can be avoided. A use of an `=delete` function cannot be compiled.
Maybe I don't know exactly what you mean by a "default unreachable implementation". This macro is a runtime failure, so that's not as good as a compile failure. But if there is actually a use of the function in some way that is implementation-dependent, then I agree we shouldn't leave it to chance.

I'm curious about how this ended up being a compiler-dependent issue, but not enough to block the PR.

Reply:
Agreed that this isn't ideal. The advantage of using the #ifdef is that our non-Windows builders continue to statically validate that there is no attempt to call this deleted function, which gives us confidence that we'll never reach this MOZ_ASSERT_UNREACHABLE() on the Windows build.

I haven't encountered any issue with this deleted function in gcc: it compiles correctly in Evergreen for me.

Unfortunately, I wasn't able to narrow this down to a simple enough case that I could definitively pinpoint the source of the problem. MSVC is more conservative about which call sites it think may be calling this function. From my inspection, I feel confident that this function cannot be called and that the compiler error is a false alarm, but I will have to go back and run the compiler again to get the exact message. (That will take me an hour or so to do, because my Windows Evergreen instance somehow became inaccessible :( ).